File: Grind75/JS2/_704_Binary_Search.js
var search = function (nums, target) {
  let start = 0,
    end = nums.length - 1;
  while (start < end) {
    let mid = Math.floor((start + end) / 2);
    if (nums[mid] === target) {
      return mid;
    } else if (nums[mid] < target) {
      start = mid + 1;
    } else {
      end = mid - 1;
    }
  }
  return -1;
};

nums = [-1, 0, 3, 5, 9, 12];
target = 9;

console.log(search(nums, target));

File: 2024/other/mockInterview/704BinarySearch.py
class Solution:
    def seach(self, nums, target):
        left, right  = 0, len(nums)-1
        while left < right:
            mid = left + (right - left)//2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1

nums = [-1,0,3,5,9,12]
target = 9

# O(logn)
# O(1)

sol = Solution()
result = sol.seach(nums, target)
print(result)